docs_src/design/adr/system-management/0006-Metrics-Collection.md

```diff
@@ -0,0 +1,275 @@
+- Count of any accesses and failures to the data persistence layer
+- Count of service start and restart attempts
+
+### Design Proposal
+
+- Each metric must have a unique name (or key) associate with it.  Because some metrics are reported from multiple services (such as service uptime), the name is not required to be unique across all services.  However, the metric name and the service name/key for a metric (see Bucket definition below) along with the origin timestamp would uniquely identify a metric from a service at a designated time.
+- All services will use/integrate go-mod-messaging (if not already integrated).
+- A new metric DTO (Telemetry/Bucket) is required for services to structure the metric data.  The model/DTO will have similarities to the core event/reading model/DTO.
+    - The DTO for the metric data will allow originating service information (the service key) to be provided in the DTO
+    - The proposed message structure (DTO) for metrics (for inclusion in go-mod-core-contracts) is:
+ 
+``` go
+type Telemetry struct {
+    Id          string            `json:"id,omitempty" codec:"id,omitempty"`  // UUID to identify the telemetry group
+    Service     string            `json:"device,omitempty" codec:"device,omitempty"`  // originating service key
+    Origin      int64             `json:"origin,omitempty" codec:"origin,omitempty"`
+    Buckets    []Bucket           `json:"readings,omitempty" codec:"buckets,omitempty"`
+    Tags        map[string]string `json:"tags,omitempty" codec:"tags,omitempty" xml:"-"`
+}
+
+type Bucket struct {
+    Id            string `json:"id,omitempty" codec:"id,omitempty"`  // UUID to identify the Bucket
+    Origin        int64  `json:"origin,omitempty" codec:"origin,omitempty"`
+    Service       string `json:"device,omitempty" codec:"device,omitempty"`   // originating service key
+    Name          string `json:"name,omitempty" codec:"name,omitempty"`  // metric name key
+    Value         string `json:"value,omitempty" codec:"value,omitempty"` // metric value
+    ValueType     string `json:"valueType,omitempty" codec:"valueType,omitempty"`  // metric value type
+}
+```
+
+!!! Note
+    Metrics are considered immutable and therefore not requiring a modified timestamp.
+
+    *For consideration, should we keep the metric data values simple (i.e. just a number field) since the metrics are just numbers and thereby avoid having to deal with types (pulling ValueType from above).  Even times could be sent as UTC number values.*
+
+    *Also for consideration, can we simplify the structures?*
+
+    - would the telemetry and bucket need an ID?  
+    - would origin timestamp need to be put on the telemetry object since it is on a Bucket?
+    - would originating service need to be put in both telemetry and bucket objects?
+    - Can the description be found in documentation and therefore left off of the Bucket?  *(@cloudxxx8 recommends removing.  Can add back in if there are dissenting opinions)*
+

```

Comment:
IMO this is re-inventing what is already available via packages like go-metrics. We should just use the raw metrics format from the chosen package.

Reply:
@lenny-intel - while I agree with your sentiment, in this case I don't know that it is practical for a couple of reasons.  The go-metrics package provides he means for collecting the telemetry data as key/values, but doesn't really provide the means to add the additional information (like service, tags, etc.) to the telemetry data when it is sent out/exported - which we would want to know who/where the telemetry came from (adding context)  Second of all, the way go-metrics works is that we are going to have to develop the means of getting metrics collected by go-metrics and then send or export them via MQTT.  So regardless of what information go-metrics has and what structure it uses (which isn't much), we are going to have to package that in an MQTT or other message bus message.  Since we are going to have to essentially pull the data, create a message, and push that message to a bus, it would seem to me that we want to create a message structure that adds the context others will need and make it look similar to our event/readings but for telemetry.

Reply:
per monthly architect's meeting - agreed to make the structure flat and allow user to format/structure the data as necessary. All that is needed is key/value pair with tags array.